src/modules/trajectory_generation/domain/spline_generation/cubic_spline_generator.cpp

```diff
@@ -1,10 +1,160 @@
 #include <cubic_spline_generator.hpp>
+#include <core_datastructures/Posture.hpp>
+#include <vector>
+#include <cmath>
+#include <eigen3/Eigen/Dense>
+#include <iostream>
+
 
 namespace trajectory_generation{
-    CubicSplineGenerator::CubicSplineGenerator(const core_datastructures::Posture& start, const core_datastructures::Posture& goal) : start(start), goal(goal) {
+
+    CubicSplineGenerator::CubicSplineGenerator(core_datastructures::Posture& start, core_datastructures::Posture& goal){
+        this->start = start;
+        this->goal = goal;
+        
+        P0 = start.kappa;
+        P3 = goal.kappa;
+        SG = get_distance(start, goal);
+
+        P1=0.1;
+        P2=0.1;
+        q_thresh(0) =0.01;
+        q_thresh(1) = 0.01;
+        q_thresh(2)=0.01;
+    }
+
+
+    std::vector<core_datastructures::Posture> CubicSplineGenerator::get_spline(){
+        run_gradient_descent();   
+        std::cout << "Final Spline Parmaeters : " << P1 << " " << P2 << std::endl;
+        return generate_splines();
+    }
+
+    std::vector<double> CubicSplineGenerator::calculate_parameters(double p1, double p2, double s){
+        double a = P0;
+        double b = -(11*P0 - 18*p1 + 9*p2 - 2*P3)/(2*s);
+        double c = 9*(2*P0 -5*p1 + 4*p2 - P3)/(2*pow(s,2));
+        double d = -9*(P0 -3*p1 + 3*p2 -P3)/(2*pow(s,3));
+
+        return {a,b,c,d};
+    }
+
+    core_datastructures::Posture CubicSplineGenerator::get_next_state(std::vector<double>& params, double s){
+        core_datastructures::Posture next_state;
+        next_state.kappa = calculate_kappa(params,s);
+        next_state.theta = calculate_theta(params,s);
+        next_state.x = calculate_x(params,s);
+        next_state.y = calculate_y(params,s);
+
+        return next_state;
+    }
+
+    double CubicSplineGenerator::calculate_kappa(std::vector<double>& params, double s){
+        return params[0] + params[1]*s + params[2]*pow(s,2) + params[3]*pow(s,3);
+    }
+
+    double CubicSplineGenerator::calculate_theta(std::vector<double>& params, double s){
+        return params[0]*s + params[1]*pow(s,2)/2 + params[2]*pow(s,3)/3 + params[3]*pow(s,4)/4;
+    }
     
+    Eigen::Matrix3d CubicSplineGenerator::calculate_Jacobian(std::vector<double>& params){
+        Eigen::Matrix3d J;
+        core_datastructures::Posture state = get_next_state(params, SG);
+        std::vector<double> values  {P1, P2, SG};
```

Comment:
A cleaner version would be to declare std::vector<double> perturb_params{} as a private variable for P1, P2, SG.
This will remove creating duplicates for P1, P2, SG during this iteration